### ISSUE TITLE
No connectivity to endpoints with systemd 242

### ISSUE THREAD

**Issue by brb** [cilium/cilium#8304]
@gandro mentioned to me that `cilium-health` with a standalone `cilium-agent` from v1.5.3 does not work on his laptop running the latest archlinux. After I upgraded my archlinux machine to the latest which upgraded Linux kernel to v5.1.6, Docker to 18.09.6-ce, systemd to 242 (242.29-1-arch). I noticed that neither `cilium-health` nor any cilium endpoint can't be reached from the host.

I've tried to curl the nginx endpoint from the host netns, and `cilium monitor` showed that a TCP SYN packet is being redirected to the peer end of the veth device in the endpoint's netns after https://github.com/cilium/cilium/blob/v1.5.3/bpf/bpf_lxc.c#L999 has been called.

The kernel stack trace showed that the packet is being dropped in `ip_rcv_core`:
```
# bpftrace -e 'kprobe:kfree_skb { @[kstack] = count(); }'
Attaching 1 probe...
^C

@[
    kfree_skb+1
    ip_rcv_core.isra.2+1025
    ip_rcv+58
    __netif_receive_skb_one_core+82
    process_backlog+155
    net_rx_action+324
    __softirqentry_text_start+274
    do_softirq_own_stack+42
    do_softirq+93
    __local_bh_enable_ip+105
    ip_finish_output2+419
    ip_output+112
    __ip_queue_xmit+354
    __tcp_transmit_skb+1398
    tcp_connect+1651
    tcp_v4_connect+1119
    __inet_stream_connect+209
    inet_stream_connect+54
    __sys_connect+135
    __x64_sys_connect+22
    do_syscall_64+91
    entry_SYSCALL_64_after_hwframe+68
]: 1
```

`perf probe` revealed that the drop happened due to `skb->pkt_type == PACKET_OTHERHOST`: https://github.com/torvalds/linux/blob/v5.1/net/ipv4/ip_input.c#L428-L429.

The `skb->pkt_type = PACKET_OTHERHOST` is set by the kernel after the redirect in bpf_lxc.c is executed. In particular, `eth_trans_type` set it for us: https://github.com/torvalds/linux/blob/v5.1/net/ethernet/eth.c#L177. The relevant ktrace:

```
@[
    eth_type_trans+1
    __dev_forward_skb+184
    veth_xmit+110
    dev_hard_start_xmit+156
    __dev_queue_xmit+2185
    __bpf_redirect+428
    __dev_queue_xmit+1864
    ip_finish_output2+661
    ip_output+112
    __ip_queue_xmit+354
    __tcp_transmit_skb+1398
    tcp_connect+1651
    tcp_v4_connect+1119
    __inet_stream_connect+209
    inet_stream_connect+54
    __sys_connect+135
    __x64_sys_connect+22
    do_syscall_64+91
    entry_SYSCALL_64_after_hwframe+68
]: 1
```

The change of the packet type was due to different dmac addrs: `skb->dev->dev_addr` was set to the veth peer in the endpoint's netns (correct), while `((struct ethhdr *)skb->data)->h_dest` was set to some mac addr which didn't exist on the system in any netns.

Next steps:

- [x] Downgrade the kernel to check whether it's a kernel regression.
- [x] Check where the non-existing mac addr comes from.

**Comment by brb** [comment]
Downgrading to 5.0.8 didn't help, so it's most likely not a kernel related change.

**Comment by borkmann** [comment]
> I noticed that neither cilium-health nor any cilium endpoint can't be reached from the host.

What about older kernels, does it work with stock Ubuntu 18.04 LTS, for example?

**Comment by brb** [comment]
Just checked on ubuntu-next (`Linux runtime1 5.1.0-rc5+`): the redirected packet has the right src and dst mac addrs: src is of the veth end in the host netns, while dst if os the veth end in the endpoint netns.

**Comment by brb** [comment]
That's where the mac addrs get modified to the invalid ones: https://github.com/cilium/cilium/blob/v1.5.3/bpf/lib/l3.h#L117.

**Comment by brb** [comment]
Seems that the veth interfaces are being managed by systemd (thus the change in their mac addrs after they have been created). Will validate my assumption tomorrow.

**Comment by borkmann** [comment]
> Seems that the veth interfaces are being managed by systemd (thus the change in their mac addrs after they have been created). Will validate my assumption tomorrow.

If that indeed holds true, I'm wondering what is the rationale for this ... they are being assigned randomly by the kernel.

**Comment by brb** [comment]
Yeah, downgrading systemd to 240 resolved the issue.

**Comment by borkmann** [comment]
Looks like it's trying to generate a persistent mac, potentially related: https://github.com/systemd/systemd/pull/11382

**Comment by brb** [comment]
TL;DR systemd-udev tries to set a persistent MAC addr for any virtual device (e.g. veth) with a default configuration (`MACAddressPolicy=persistent`). This breaks Cilium, as MAC addrs are being set to different ones for an endpoint veth pair than it had in the beginning, so `cilium_lxc` ends up having stale MAC addrs which makes the Linux kernel to drop packets sent to the endpoint. The PR which introduced this behaviour: https://github.com/systemd/systemd/pull/11382.

Possible workarounds:

1. Explicitly set MAC addrs for the veth pair when creating it via `netlink.LinkAdd()`. This should prevent systemd-udev from changing the addrs: https://github.com/systemd/systemd/blob/v242/src/udev/net/link-config.c#L292.
2. Change `MACAddressPolicy` to `none`. Cons: should be done by a user.
3. Monitor MAC changes via `netlink.SubscribeLinks()` and reflect them in them `cilium_lxc` map. Cons: small windows of unavailability; increased complexity.

**Comment by borkmann** [comment]
> 1. Explicitly set MAC addrs for the veth pair when creating it via `netlink.LinkAdd()`. This should prevent systemd-udev from changing the addrs: https://github.com/systemd/systemd/blob/v242/src/udev/net/link-config.c#L292.

We should go for this one iff feasible so that users don't have to manually work around it.

**Comment by nirmoy** [comment]
How about retrieving kernel generated mac after `netlink.LinkAdd` and the calling `netlink.LinkSetHardwareAddr` so that we don't have to create a separate logic to generate Mac address

**Comment by brb** [comment]
> How about retrieving kernel generated mac after netlink.LinkAdd and the calling netlink.LinkSetHardwareAddr so that we don't have to create a separate logic to generate Mac address

I think with this approach it's possible for the following race to happen:

1. cilium: `netlink.LinkAdd()`.
2. cilium: `netlink.LinkByName()`.
3. udev: Get `addr_assign_type` (=`NET_ADDR_RANDOM`).
4. cilium: `netlink.LinkSetHardwareAddr()`.
5. udev: Set mac.

**Comment by nirmoy** [comment]
@brb agreed

**Comment by brb** [comment]
> Is this problem only for veth or IPVLAN as well ?

Should be affected both.

Hmm, it's not possible to specify MAC addr of a veth peer when creating it via `netlink.LinkAdd()`. I'm going to open a PR to `vishvananda/netlink` first.

**Comment by nirmoy** [comment]
@brb we can at least configure mac for one host side 
check https://play.golang.org/p/S2wY3Hq4Ick

**Comment by brb** [comment]
> we can at least configure mac

Yep, I know. Unfortunately, both ends need to be configured.

**Comment by brb** [comment]
netlink PR: https://github.com/vishvananda/netlink/pull/460.
